docs/proposals/post-deployment-hooks.md

```diff
@@ -0,0 +1,325 @@
+2. Hook execution can't be atomically linked to the deployment pods' statuses: If a hook failure should result in deployment failure, a previously scaled and exposed application must be rolled back when ideally the application shouldn't have been exposed prior to hook success.
+
+
+## Use cases
+
+1. As a Rails application developer, I want to perform a Rails database migration following an application deployment.
+2. As a Java application developer, I want to invoke a cloud API endpoint to notify it of the presence of new code.
+
+
+#### Use-case: Rails migrations
+
+New revisions of a Rails application often contain schema or other database updates which must accompany the new code deployment. Users should be able to specify a hook which performs a Rails migration as part of the application code deployment.
+
+Database migrations are complex and introduce downtime concerns. Here are [some examples](https://blog.rainforestqa.com/2014-06-27-zero-downtime-database-migrations) of zero-downtime Rails migration workflows.
+
+Deployments including database migrations must make special considerations:
+
+1.  Code must be newer than the in use schema, or all old code must be stopped before the new schema is introduced.
+2.  Database or table locking must be minimized or eliminated to prevent service outages.
+
+The workflows which are effective at ensuring zero downtime migrations are typically multi-phased. For a user orchestrating a zero downtime migration deployment, it's likely the user needs to verify each deployment step discretely, with the option to abort and rollback after each phase.
+
+###### Container lifecycle hooks
+
+Container lifecycle hooks introduce problems with Rails migrations:
+
+1. There is no way to guarantee that pods with older code are not running.
+2. The migration hook will execute in the same pod as application containers, consuming resources allocated for the application.
+  1. This can cause instability, as it's unlikely the application pod resource allocation takes into account the temporarily increased requirements of a transient deployment step.
+
+###### Deployment hooks
+
+Deployment hooks satisfy this use case by providing a means to execute the hook only once per logical deployment. The hook is expressed as a run-once pod which provides the migration with its own resource allocation decoupled from the application.
+
+
+#### Use-case: Invoke a cloud API endpoint
+
+Consider a Java application whose deployment should result in a cloud API call being invoked to notify it of the newly deployed code.
+
+###### Container lifecycle hooks
+
+Container lifecycle hooks aren't ideal for this use case because they will be fired once per pod in the deployment during scale-up rather than following the logical deployment as a whole.
+
+###### Deployment hooks
+
+A post-deployment hook would satisfy the use case by ensuring the API call is invoked after the deployment has been rolled out.
```

Comment:
This is another area where a little example of the API would be great.  It doesn't have to match the API that you eventually propose.